src/Lucene.Net/Index/SegmentReader.cs

```diff
@@ -191,10 +191,16 @@ internal Directory GetCFSReader()
 					return cfsReader;
 				}
 			}
-			
+
 			internal TermInfosReader GetTermsReader()
-			{
-				lock (this)
+            {
+                // This is initialized in the constructor.
+                // The only time it isn't initialized is when termsIndexDivisor equals -1, which we don't use.
+                var reader = tis;

```

Comment:
What if `tis` will be already disposed? As I see it can be disposed by `DecRef()` and since here we're no longer under the lock that seems to be possible

Reply:
We can still get a reference to it and after that it will be disposed.
There is no protection against it in the current implementation.

Reply:
In current impl if `tis` is `null` (which is the case after it got disposed: https://github.com/ravendb/lucenenet/blob/34bc333c964b47e4c8069e726ba1d316d07eb664/src/Lucene.Net/Index/SegmentReader.cs#L261-L266)

then we return `tisNoIndex`.

After the change made here we can return `tis` while `DecRef()` will dispose it (don't know if those can be executed concurrently but I believe they can since both use `lock()`)

Reply:
You can get a reference to it (when calling `GetTermsReader`).
And after that it can be disposed, so you'll be holding a disposed object.

Reply:
The original code doesn't protect you from getting a `tis` this might be disposed afterwards.